File: primitives/fee_market.go
package primitives

import "math"

// EIP-1559 Fee Market implementation
// Base fee adjusts based on block fullness; burned, not paid to miners

// MinBaseFee is the minimum base fee in wei
const MinBaseFee uint64 = 7

// calculateFeeDelta safely calculates fee delta avoiding overflow and division by zero
func calculateFeeDelta(fee, gasDelta, gasTarget, denominator uint64) uint64 {
	// Use larger precision for intermediate calculation to avoid overflow
	// Convert to uint128-equivalent using Go's native uint64 arithmetic with checks

	// Avoid division by zero
	if gasTarget == 0 || denominator == 0 {
		divisor := uint64(1)
		// fee * gasDelta with overflow protection
		if gasDelta == 0 || fee == 0 {
			return max(1, 0)
		}
		// Check for overflow: if fee * gasDelta would overflow
		if gasDelta > math.MaxUint64/fee {
			return math.MaxUint64
		}
		intermediate := fee * gasDelta
		return max(1, intermediate/divisor)
	}

	divisor := gasTarget * denominator

	// Check for overflow in multiplication
	if gasDelta > 0 && fee > math.MaxUint64/gasDelta {
		// Would overflow, return max safe value
		return max(1, math.MaxUint64/divisor)
	}

	intermediate := fee * gasDelta
	result := intermediate / divisor

	// Always return at least 1 to ensure some movement
	return max(1, result)
}

// max returns the maximum of two uint64 values
func max(a, b uint64) uint64 {
	if a > b {
		return a
	}
	return b
}

// min returns the minimum of two uint64 values
func min(a, b uint64) uint64 {
	if a < b {
		return a
	}
	return b
}

// InitialBaseFee calculates the base fee for the first EIP-1559 block based on parent gas usage
//
// Parameters:
//   - parentGasUsed: Gas used by the parent block
//   - parentGasLimit: Gas limit of the parent block
//
// Returns:
//   - Initial base fee in wei
func InitialBaseFee(parentGasUsed, parentGasLimit uint64) uint64 {
	parentGasTarget := parentGasLimit / 2
	baseFee := BaseFeeInitial // 1 gwei

	if parentGasUsed > 0 {
		var gasUsedDelta uint64
		if parentGasUsed > parentGasTarget {
			gasUsedDelta = parentGasUsed - parentGasTarget
		} else {
			gasUsedDelta = parentGasTarget - parentGasUsed
		}

		baseFeeDelta := calculateFeeDelta(baseFee, gasUsedDelta, parentGasTarget, BaseFeeMaxChangeDenominator)

		if parentGasUsed > parentGasTarget {
			baseFee = baseFee + baseFeeDelta
		} else if baseFee > baseFeeDelta {
			baseFee = baseFee - baseFeeDelta
		}
	}

	return max(baseFee, MinBaseFee)
}

// NextBaseFee calculates the next block's base fee based on the current block
//
// This implements the EIP-1559 base fee adjustment algorithm:
// - If the block used exactly the target gas, the base fee remains the same
// - If the block used more than the target, the base fee increases
// - If the block used less than the target, the base fee decreases
// - The maximum change per block is 12.5% (1/8)
//
// Parameters:
//   - parentBaseFee: Base fee of the parent block (in wei)
//   - parentGasUsed: Gas used by the parent block
//   - parentGasTarget: Target gas usage of the parent block
//
// Returns:
//   - The next block's base fee (in wei)
func NextBaseFee(parentBaseFee, parentGasUsed, parentGasTarget uint64) uint64 {
	// If parent block is empty, keep the base fee the same
	if parentGasUsed == 0 {
		return parentBaseFee
	}

	newBaseFee := parentBaseFee

	if parentGasUsed == parentGasTarget {
		// If parent block used exactly the target gas, keep the base fee the same
		// No change needed
	} else if parentGasUsed > parentGasTarget {
		// If parent block used more than the target gas, increase the base fee
		gasUsedDelta := parentGasUsed - parentGasTarget

		// Calculate the base fee delta (max 12.5% increase)
		baseFeeDelta := calculateFeeDelta(parentBaseFee, gasUsedDelta, parentGasTarget, BaseFeeMaxChangeDenominator)

		// Increase the base fee with overflow protection
		if parentBaseFee > math.MaxUint64-baseFeeDelta {
			// Would overflow, keep at parent fee
			newBaseFee = parentBaseFee
		} else {
			newBaseFee = parentBaseFee + baseFeeDelta
		}
	} else {
		// If parent block used less than the target gas, decrease the base fee
		gasUsedDelta := parentGasTarget - parentGasUsed

		// Calculate the base fee delta (max 12.5% decrease)
		baseFeeDelta := calculateFeeDelta(parentBaseFee, gasUsedDelta, parentGasTarget, BaseFeeMaxChangeDenominator)

		// Decrease the base fee, but don't go below the minimum
		if parentBaseFee > baseFeeDelta {
			newBaseFee = parentBaseFee - baseFeeDelta
		} else {
			newBaseFee = MinBaseFee
		}
	}

	// Ensure base fee is at least the minimum
	return max(newBaseFee, MinBaseFee)
}

// EffectiveGasPriceResult contains the results of effective gas price calculation
type EffectiveGasPriceResult struct {
	EffectiveGasPrice uint64 // The effective gas price paid
	MinerFee          uint64 // The miner's portion of the fee (priority fee)
}

// GetEffectiveGasPrice calculates the effective gas price for an EIP-1559 transaction
//
// The effective gas price is the minimum of:
//  1. maxFeePerGas specified by the sender
//  2. baseFeePerGas + maxPriorityFeePerGas
//
// Parameters:
//   - baseFeePerGas: Current block's base fee
//   - maxFeePerGas: Maximum fee the sender is willing to pay
//   - maxPriorityFeePerGas: Maximum tip the sender is willing to pay to the miner
//
// Returns:
//   - The effective gas price, and the miner's portion of the fee
func GetEffectiveGasPrice(baseFeePerGas, maxFeePerGas, maxPriorityFeePerGas uint64) EffectiveGasPriceResult {
	// Ensure the transaction at least pays the base fee
	if maxFeePerGas < baseFeePerGas {
		// Transaction's max fee is less than base fee
		// In a real implementation, this transaction would be rejected
		// For now, just return the max fee and zero miner fee
		return EffectiveGasPriceResult{
			EffectiveGasPrice: maxFeePerGas,
			MinerFee:          0,
		}
	}

	// Calculate the priority fee (tip to miner)
	// This is limited by both maxPriorityFeePerGas and the leftover after base fee
	maxPriorityFee := min(maxPriorityFeePerGas, maxFeePerGas-baseFeePerGas)

	// The effective gas price is base fee plus priority fee
	effectiveGasPrice := baseFeePerGas + maxPriorityFee

	return EffectiveGasPriceResult{
		EffectiveGasPrice: effectiveGasPrice,
		MinerFee:          maxPriorityFee,
	}
}

// GetGasTarget calculates the gas target for a block
//
// The gas target is the desired gas usage per block, which is typically
// half of the maximum gas limit.
//
// Parameters:
//   - gasLimit: The block's gas limit
//
// Returns:
//   - The gas target for the block
func GetGasTarget(gasLimit uint64) uint64 {
	return gasLimit / 2
}

// CalculateNextBaseFee is a convenience function that calculates the next base fee
// given a parent gas limit instead of a pre-calculated gas target
//
// Parameters:
//   - parentBaseFee: Base fee of the parent block (in wei)
//   - parentGasUsed: Gas used by the parent block
//   - parentGasLimit: Gas limit of the parent block
//
// Returns:
//   - The next block's base fee (in wei)
func CalculateNextBaseFee(parentBaseFee, parentGasUsed, parentGasLimit uint64) uint64 {
	parentGasTarget := GetGasTarget(parentGasLimit)
	return NextBaseFee(parentBaseFee, parentGasUsed, parentGasTarget)
}

// CalculatePriorityFee calculates the priority fee (miner tip) for a transaction
//
// Parameters:
//   - maxFeePerGas: Maximum fee per gas willing to pay
//   - baseFee: Current block's base fee
//   - maxPriorityFeePerGas: Maximum priority fee per gas
//
// Returns:
//   - Actual priority fee per gas
func CalculatePriorityFee(maxFeePerGas, baseFee, maxPriorityFeePerGas uint64) uint64 {
	if maxFeePerGas < baseFee {
		return 0
	}

	maxPriorityFee := maxFeePerGas - baseFee
	return min(maxPriorityFee, maxPriorityFeePerGas)
}

File: src/primitives/v2/fee-market.ts
// EIP-1559 & EIP-4844 Fee Market Implementation
// References:
// - EIP-1559: https://eips.ethereum.org/EIPS/eip-1559
// - EIP-4844: https://eips.ethereum.org/EIPS/eip-4844

// EIP-1559 Constants
export const MIN_BASE_FEE = 7n; // Minimum base fee (wei)
export const BASE_FEE_CHANGE_DENOMINATOR = 8n; // Max 12.5% change per block

// EIP-4844 Constants
export const MIN_BLOB_BASE_FEE = 1n; // Minimum blob base fee (wei)
export const BLOB_BASE_FEE_UPDATE_FRACTION = 3338477n; // Exponential update denominator
export const TARGET_BLOB_GAS_PER_BLOCK = 393216n; // 3 blobs worth
export const BLOB_GAS_PER_BLOB = 131072n; // 2^17

/**
 * Calculate next block's base fee using EIP-1559 formula
 *
 * Formula:
 * - If parentGasUsed > parentGasTarget:
 *   baseFee = parentBaseFee + parentBaseFee * (parentGasUsed - parentGasTarget) / parentGasTarget / 8
 * - If parentGasUsed < parentGasTarget:
 *   baseFee = parentBaseFee - parentBaseFee * (parentGasTarget - parentGasUsed) / parentGasTarget / 8
 * - Else: baseFee = parentBaseFee
 * - Always: baseFee >= MIN_BASE_FEE
 *
 * @param parentGasUsed - Gas used in parent block
 * @param parentGasLimit - Gas limit of parent block
 * @param parentBaseFee - Base fee of parent block (wei)
 * @returns Next block's base fee (wei)
 */
export const calculateBaseFee = (
  parentGasUsed: bigint,
  parentGasLimit: bigint,
  parentBaseFee: bigint,
): bigint => {
  // Target is 50% of gas limit (elasticity multiplier = 2)
  const parentGasTarget = parentGasLimit / 2n;

  // Empty block: base fee stays same
  if (parentGasUsed === 0n) {
    return parentBaseFee;
  }

  // At target: base fee stays same
  if (parentGasUsed === parentGasTarget) {
    return parentBaseFee;
  }

  let newBaseFee = parentBaseFee;

  if (parentGasUsed > parentGasTarget) {
    // Block used more than target: increase base fee
    const gasUsedDelta = parentGasUsed - parentGasTarget;
    // baseFee * delta / target / 8 (max 12.5% increase)
    const baseFeeDelta =
      (parentBaseFee * gasUsedDelta) / parentGasTarget / BASE_FEE_CHANGE_DENOMINATOR;
    newBaseFee = parentBaseFee + (baseFeeDelta > 0n ? baseFeeDelta : 1n);
  } else {
    // Block used less than target: decrease base fee
    const gasUsedDelta = parentGasTarget - parentGasUsed;
    // baseFee * delta / target / 8 (max 12.5% decrease)
    const baseFeeDelta =
      (parentBaseFee * gasUsedDelta) / parentGasTarget / BASE_FEE_CHANGE_DENOMINATOR;
    const delta = baseFeeDelta > 0n ? baseFeeDelta : 1n;
    newBaseFee = parentBaseFee > delta ? parentBaseFee - delta : MIN_BASE_FEE;
  }

  // Enforce minimum base fee
  return newBaseFee < MIN_BASE_FEE ? MIN_BASE_FEE : newBaseFee;
};

/**
 * Calculate blob base fee using EIP-4844 formula
 *
 * Formula: fakeExponential(MIN_BLOB_BASE_FEE, excessBlobGas, BLOB_BASE_FEE_UPDATE_FRACTION)
 *
 * fakeExponential approximates: factor * e^(numerator / denominator)
 * Using Taylor series: factor * (1 + x + x^2/2 + x^3/6 + ...)
 * where x = numerator / denominator
 *
 * @param excessBlobGas - Excess blob gas from previous blocks
 * @returns Blob base fee (wei)
 */
export const calculateBlobBaseFee = (excessBlobGas: bigint): bigint => {
  // fakeExponential(MIN_BLOB_BASE_FEE, excessBlobGas, BLOB_BASE_FEE_UPDATE_FRACTION)
  return fakeExponential(MIN_BLOB_BASE_FEE, excessBlobGas, BLOB_BASE_FEE_UPDATE_FRACTION);
};

/**
 * Calculate excess blob gas for next block using EIP-4844 formula
 *
 * Formula:
 * excessBlobGas = max(0, parentExcessBlobGas + parentBlobGasUsed - TARGET_BLOB_GAS_PER_BLOCK)
 *
 * @param parentExcessBlobGas - Excess blob gas from parent block
 * @param parentBlobGasUsed - Blob gas used in parent block
 * @returns Excess blob gas for next block
 */
export const calculateExcessBlobGas = (
  parentExcessBlobGas: bigint,
  parentBlobGasUsed: bigint,
): bigint => {
  const total = parentExcessBlobGas + parentBlobGasUsed;

  if (total < TARGET_BLOB_GAS_PER_BLOCK) {
    return 0n;
  }

  return total - TARGET_BLOB_GAS_PER_BLOCK;
};

/**
 * Approximate exponential function using Taylor series
 * Internal helper for EIP-4844 blob fee calculation
 *
 * Approximates: factor * e^(numerator / denominator)
 *
 * @param factor - Base factor (MIN_BLOB_BASE_FEE)
 * @param numerator - Exponent numerator (excessBlobGas)
 * @param denominator - Exponent denominator (BLOB_BASE_FEE_UPDATE_FRACTION)
 * @returns Approximated result
 */
const fakeExponential = (factor: bigint, numerator: bigint, denominator: bigint): bigint => {
  let output = 0n;
  let numeratorAccum = factor * denominator;
  let i = 1n;

  // Taylor series: sum of (numerator^i) / (denominator^i * i!)
  while (numeratorAccum > 0n && i <= 256n) {
    output += numeratorAccum;
    numeratorAccum = (numeratorAccum * numerator) / (denominator * i);
    i += 1n;
  }

  return output / denominator;
};

export const FeeMarket = {
  calculateBaseFee,
  calculateBlobBaseFee,
  calculateExcessBlobGas,
} as const;
